### ISSUE TITLE
Feature Requests and Ideas

### ISSUE THREAD

**Issue by sebastianwessel** [sebastianwessel/quickjs#10]
If you have ideas, wishes, feature request and feedback - please add them in the comments below

**Comment by poef** [comment]
Can I also inject javascript objects from the main process into the QuickJS runtime? I'm building a node express based data server, where I want to allow clients to run javascript queries on that data (https://github.com/SimplyEdit/SimplyStore/)

**Comment by sebastianwessel** [comment]
Hey @poef 

Yes, you can. See [Data Exchange Between Host and Guest](https://sebastianwessel.github.io/quickjs/basic.html)
You can use `env`, and provide strings, numbers, arrays, objects and functions.

```typescript
const { evalCode } = await createRuntime({
  env: {
    MY_PROCESS_ENV: 'some environment variable provided by the host',
    KV: {
      set: (key: string, value: string) => keyValueStoreOnHost.set(key, value),
      get: (key: string) => keyValueStoreOnHost.get(key),
    },
  },
})
```

**Comment by guest271314** [comment]
Read standard input to V8's `d8` (`/proc/PID/fd/0`) with WebAssembly. Right now I'm using QuickJS via `os.system()` https://github.com/guest271314/native-messaging-d8/blob/quickjs-stdin-read/nm_d8.js#L16, https://github.com/guest271314/native-messaging-d8/blob/quickjs-stdin-read/read_d8_stdin.js. If we can do this with WebAssembly we can get rid of `os.system()` which call `sh`.

**Comment by sebastianwessel** [comment]
@guest271314 thanks for your feedback. As far as I understand, if the "regular" deadline nodejs module is available inside of the quickjs runtime, it should solve the issue or?
So, the code inside the sandbox would look like this

'''Javascript

const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

rl.on('line', (line) => {
    console.log(line);
});

rl.once('close', () => {
     // end of input
 });
'''

**Comment by guest271314** [comment]
> As far as I understand, if the "regular" deadline nodejs module is available inside of the quickjs runtime

Where does Node.js API's get in to your WASM QuickJS build?

The idea is to use the least amount of resources to read `d8` shell standard input. QuickJS (quickjs-ng) is around 1.2 MB. 

I was thinking I could use WebAssembly/WASI to read stdin to `d8` using `WebAssembly.compile()`.

**Comment by guest271314** [comment]
I'm basically trying to do this https://github.com/guest271314/native-messaging-webassembly without a WASM runtime, using the built-in `WebAssembly` object alone - within `d8`. I created a different solution for Mozilla SpiderMonkey.

**Comment by sebastianwessel** [comment]
What I was meaning is, that you can provide functions and data from a regular Node.js host application, to the wasm guest system.
Basically, you would read the standard-in in the host application and provide this data than to the guest system, as a copy.
What you try is, to disable the isolation of the wasm, and to allow to access the host system directly. Kind of disable security and enhance WASI in the direction of wasix.
In this project, this not wanted, as the JavaScript should run in an isolated sandbox, ensuring that it is not possible to break out, and access host functionality directly, in an uncontrolled manner.

**Comment by guest271314** [comment]
> What I was meaning is, that you can provide functions and data from a regular Node.js host application, to the wasm guest system.

It doesn't make sense to me to use 108.7 MB `node` executable that depends on V8 to read standard input to `v8` at 37.9 MB.

I use `deno` and `bun` and `qjs` and `tjs` the same that I use `node`, so I don't think of `node` as a "regular Node.js application". `node` is just another JavaScript runtime in the JavaScript toolbox for me.

That's why I chose `qjs` at 1.2 MB to do the task.

I have been trying to do this using `d8`s `readline()`, though have not succeeded, yet.

I saw your work and this issue requesting features and decided to place a feature request.

> In this project, this not wanted, as the JavaScript should run in an isolated sandbox, ensuring that it is not possible to break out, and access host functionality directly, in an uncontrolled manner.

I don't think it is possible to achieve that requirement. I have broken out of too many alleged "sandbox" to think for a moment that it can't be done in this case, too. 

Thanks!

**Comment by sebastianwessel** [comment]
Hey,
No worries, it's totally wanted to place such issues - even if I can't help here.
Can ask you, what's the general use case you like to achieve?

**Comment by guest271314** [comment]
I don't have a use case for running applications in a "sandbox".  

I generally break out of sandboxes that folks try to set up.

For people interested in "sandbox" code we already have that with `Worker` and WebAssembly in and out of the browser and `SharedWorker`, `Worklet` interfaces in the browser.

**Comment by digipigeon** [comment]
It would be good to have an simple/intuative way to set limits, such as.

1. Set Limits
Maximum Memory
Max CPU Time - as a decimal between 0-1 1 being its allowed to consume 100% of the core/process that it is running on.

2. Observability
It should be possible to poll the sandbox to find out, how much memory is being consumed (either as a value or as a percentage), same as CPU time.

3. Informative Errors
If the sandbox is destroyed because it exceeds the max memory or the cpu is used too much there should be a clean informative error thrown (according to the new design it might be good to throw this at both the `runtime.runSandboxed` level as well as at the `sandbox.evalCode` level.

**Comment by sebastianwessel** [comment]
Hey @digipigeon 
thanks for your feedback. There was a similar question recently [Limit CPU and memory usage](https://github.com/sebastianwessel/quickjs/issues/35).
The idea of polling from outside is interesting, but there is no simple working solution for it in node. As long as the eval function is running, the host side is kind of blocked. If there is no eval executed, it is already possible to get memory information.
As the intention is, to have a highly controlled and isolated sandbox, even if it would be technically possible to allow direct net access, I don't think I will add it (at least per default).
The current focus is, to provide an environment, which is as close as possible to node and similar runtimes.

**Comment by digipigeon** [comment]
Hi @sebastianwessel, would you consider implementing https://nodejs.org/api/worker_threads.html#performanceeventlooputilizationutilization1-utilization2 as an alternative to accomplish something similar?

**Comment by sebastianwessel** [comment]
@digipigeon Not directly in this library, as the focus is on providing a sandbox, data (de-)serialization, runtime compatibility inside the sandbox, and developer experience (DX). 

The developer should be free to choose the method that fits best. My recommendation is to use libraries that are specialized for this, such as the [poolifier-web-worker](https://github.com/poolifier/poolifier-web-worker) package, which is used in the [Server Example here](https://github.com/sebastianwessel/quickjs/tree/main/example/server).

For instance, the usage in the browser might differ from that in the backend. In the browser, you might need only one sandbox, while in the backend, as many sandboxes as possible are required.

**Comment by aashutoshrathi** [comment]
Can there be `fuel metering` like we have in [wasmtime](https://github.com/bytecodealliance/wasmtime/issues/4109)?

**Comment by aashutoshrathi** [comment]
Can there be `fuel metering` like in [wasmtime](https://github.com/bytecodealliance/wasmtime/issues/4109)?
Since it's similar concept in terms of running JS in an isolated environment using some engine compiled to WASM

**Comment by sebastianwessel** [comment]
@aashutoshrathi as quickly is used, there is the option to do something like this:

'''typescript
let interruptCycles = 0
runtime.setInterruptHandler(() => ++interruptCycles > 1024)
'''

**Comment by aashutoshrathi** [comment]
@sebastianwessel  and I can use `interruptCycles` as fuel here?

**Comment by sebastianwessel** [comment]
Kind of - it highly depends on your use case and what you like to achieve I guess.

Personally, I do not see any real world use case, where it makes sense, to count such things, because in this case, you will need up front what is executed in the sandbox, to find the correct value.
It is simply the wrong layer to control such things imo.
When it comes to resource consumption, you probably would need to do it on the webassembly level. Meaning you would need to configure node/bun/browser to restrict the wasm resources.

**Comment by bitnom** [comment]
Please make this possible: https://github.com/justjake/quickjs-emscripten/tree/main/examples/cloudflare-workers

**Comment by guest271314** [comment]
@bitnom Interesting. Cloudflare Workers used as a server?

I use Cloaudflare `workerd`, QuickJS-NG, and Bytecode Alliance's `javy` separately.

Javy depends on QuickJS, compiles JavaScript to WASM.

The Cap 'n Proto configuration model of Cloudflare, in my opinion, is unnecessarily complicated when JSON would suffice.

Since Javy support WASI, I think in theory you should be able to use the socket interface.

**Comment by thorn0** [comment]
Did you consider extracting fs, test runner, TS support, etc., into a separate package to keep the core library lighter and more focused?

**Comment by sebastianwessel** [comment]
> Did you consider extracting fs, test runner, TS support, etc., into a separate package to keep the core library lighter and more focused?

No. At the moment, I’m focusing more on stabilization than optimization. I think the first steps are to separate synchronous and asynchronous processes and (maybe) introduce some kind of dynamic Node module (import from 'node:***') selection.

Regarding size: I guess the required duplication for sync/async handling and Node compatibility modules is much more significant than the few extra lines needed for TypeScript support or the test runner.

Maybe some kind of config and build script could be useful — so that you can selectively include only what you really need and generate a custom, optimized build. That might be the best approach 🤷‍♂️?

**Comment by guest271314** [comment]
Re TypeScript support in Node.js `node` depends on Node.js Amaro, which depends on SWC. Basically the types are just stripped.

**Comment by sebastianwessel** [comment]
@guest271314 
Think we are good with regular typescript package, as it works in browsers as well. It is also an optional dependency, and not shipped per default

**Comment by guest271314** [comment]
Why use Microsoft TypeScript `tsc` if it's not needed? `bun` doesn't use `tsc` to parse TypeScript syntax. Nor does Facebook's Static Hermes, which also can be compiled to WASM. I don't think Deno does either, except for `deno check`. You could just strip types of TypeScript input like JavaScript runtimes do.

**Comment by sebastianwessel** [comment]
@guest271314 well, because it is simple, functional, and always 100% in sync with the latest TypeScript version.

For example, if you have a TypeScript enum in your code, it won’t work in current Node versions, as TypeScript sometimes requires more than just stripping away types.

What is the issue with regular `typescript` package? In most cases - especially in the backend - it is already in the (dev)dependencies

**Comment by guest271314** [comment]
Not so long ago Microsoft TypeScript lagged to write out its version of resizable `ArrayBuffer` that had been already shipped in the browser, Node.js, and Bun.

It's a *dependency*.

TypeScript ultimately gets executed by JavaScript runtimes after stripping or "transforming" types.

Bun can execute WASM directly. Deno and Bun can execute AssemblyScript directly.

If the idea is a "sandbox" where TypeScript and JavaScript can be executed, Deno already provides that with the builtin permissions policies and flags. 

 To me the idea of using QuickJS is to get away from all kinds of dependencies.

**Comment by guest271314** [comment]
Anyway, good luck!

**Comment by thorn0** [comment]
@sebastianwessel Got it, makes sense that stabilization is the priority right now. My interest isn’t really about size or Node compatibility, but more about having a solid single-responsibility core library, namely a JS sandbox with an event loop. Extra features like Node compatibility or a test runner could then live as separate packages (wrappers, plugins, etc).

**Comment by sebastianwessel** [comment]
The issue with the event loop is, that it must be inside of the wasm. I believe, that an (fake) event loop on host side is the wrong place, and will never become a stable, solid solution.

From a host perspective, the execution of the wasm is sync (blocking). This limitation/behavior is on top of the other restricted possibilities, to control something inside the sandbox from the host. Here we hard rely on the implementation of QuickJS itself. Similar for streaming, networking access...

As an example: Something "simple" like execution timeout is simply not possible from host, as there is no "kill wasm" in node. Also, adding a logic which could trigger this kill switch is hard to implement, when the wasm execution is blocking the host event loop.

There are other interesting projects, like winterjs & co, which implemented more or less the full nodejs functionality.

**Comment by sebastianwessel** [comment]
As an example:

It would be cool if you could start a sandbox, which reads a stream from the host, and stops when the stream gets closed or the host triggers a stop.

But what you need to do is, to read the stream in the host, run the (new) sandbox for the single chunk, the sandbox finishes and returns a result, you process the result, and than you do the same again, over and over again.

And you need a fresh sandbox, as there is no garbage collection.

So, it really depends on the use case, if QuickJS is the right solution. Also, the implementation might need to be different, compared to a pure node/bun/deno implementation, to fit into the limitations.

**Comment by MulverineX** [comment]
I'd like to run web imports, eg.
```ts
 import youtubeiJs from 'https://cdn.jsdelivr.net/npm/youtubei.js@13.1.0/+esm'
```

Would it be possible for this to be implemented?

**Comment by MulverineX** [comment]
After tinkering around some, it looks like some WinterGC features are missing, the key ones being `atob`, `btoa`, and SubtleCrypto's `digest`.

For now, base64 can be implemented in pure JS, here's an example https://github.com/saghul/txiki.js/blob/master/src/js/polyfills/abba.js

And a WIP SubtleCrypto implementation on top of QuickJS does finally exist! As of 3 months ago https://github.com/awslabs/llrt/pull/698 was merged bringing support for most of it, including `digest`.

**Comment by sebastianwessel** [comment]
> I'd like to run web imports, eg.
> ```ts
>  import youtubeiJs from 'https://cdn.jsdelivr.net/npm/youtubei.js@13.1.0/bundle/cf-worker.js'
> ```
> 
> Would it be possible for this to be implemented?

Yes, with the asynchronous version, this already possible. There is also some example in the repo - here react is loaded from esm.sh

**Comment by MulverineX** [comment]
It'd be nice if the Playground had support for async mode

**Comment by sebastianwessel** [comment]
> After tinkering around some, it looks like some WinterGC features are missing, the key ones being Events (`Event`, `CustomEvent`, `EventTarget`), `atob`, `btoa`, and SubtleCrypto's `digest`.
> 
> For now, base64 can be implemented in pure JS, here's an example https://github.com/saghul/txiki.js/blob/master/src/js/polyfills/abba.js
> 
> Same goes for Events,
> https://github.com/saghul/txiki.js/blob/master/src/js/polyfills/event-target.js
> 
> And a WIP SubtleCrypto implementation on top of QuickJS does finally exist! As of 3 months ago https://github.com/awslabs/llrt/pull/698 was merged bringing support for most of it, including `digest`.
> 
> I'm very curious to know whether implementing these are in-scope for this project.

Thanks for the links.

In general, we need to differentiate between features implemented in QuickJS and missing packages. QuickJs features will more or less become available, when the wasm builds are updated. Meaning nothing needs to change in this lib.

Getting more and better standard libs and features into this lib, which are not available in plain QuickJS, is one of the key things of my lib. I will try to do my best, to add this continuously. But there are definitely some limits. Mainly when they require a real event-loop or access to underlaying OS, like network, hardware acceleration etc, which is simply not accessible from a wasm module.

**Comment by smerickson** [comment]
This library is INCREDIBLE!!  Thanks so much for making and sharing it! I'm very excited to use it in order to safely execute AI generated code!  

I'm having trouble getting the async server-side react rendering example to work.  I'm not able to load the webassembly. (see screenshot).  I'm using the exact code copy and pasted from the SSR example on the webiste (https://sebastianwessel.github.io/quickjs/use-cases/serverside-rendering.html). The only two modifications I made were

1. I ran `bun add quickjs-emscripten-core`
2. I modified to source code to import from `@sebastianwessel/quickjs` intead of `../../src/index.js`

![Image](https://github.com/user-attachments/assets/95d4e5d6-2a70-4eee-a7a0-45383c40881a)

any suggestions on other things to try in order to troubleshoot?

Also, my ultimate goal for this example would be to accept a JSX string that was generated by AI and then transpile it dynamically before then rendering it to a string.  Any suggestions on how to add the transpile step into that example?

**Comment by sebastianwessel** [comment]
Hi,
Think you do not need to add the emscripten-core explicit. It is already in the deps of this package. Only the package containing your preferred wasm file needs to be added.

Adding cache can be done. I didn't add it, as this highly depends on the use cases and environment.

Basically, you need to extend the module loader:

https://sebastianwessel.github.io/quickjs/docs/module-resolution/async-module-loader.html#🔹-step-2-fetching-modules-asynchronously

It's up to you, how you like to cage it. Maybe just in memory with as simple Map, or on the file system, or in browsers db...

Than, the module loader can be set in the options.

This is quite powerful, as you can add caching, or add more security restrictions, or change the location where to fetch the packages from.

Regarding react. I also tried the newer version, but I failed for same reasons. I didn't had the time to investigate what exactly the issue is, and how to solve it.
But, it is something I will try to investigate, as soon as I've some time. If someone else likes to figure it out... 🤷‍♂️

**Comment by smerickson** [comment]
Thanks for following up!  I was able to get my script all working the way I want!  I decided to just pre-bundle react and reactdom and dependencies into a single file and then directly include in quickjs instead of fetching them from esm.sh.  Its all working beautifully.  Thanks for this library.  Its incredible and opens up so many possibilities; especially when you pair with an LLM for dynanmicly generating code.

**Comment by gh-5000** [comment]
👋 

I recently moved off of isolated-vm, one thing I am missing is metrics about one singular execution of code.

like:
```javascript
{
  cpuTime: '586583',
  wallTime: '630751',
  mem: {
    used_heap_size: 445704,
    malloced_memory: 73800,
    peak_malloced_memory: 372736,
    total_heap_size: 1884160,
    total_heap_size_executable: 0,
    total_physical_size: 1622016
  }
}
```

Not sure if this is already possible :)
